renderlib/ViewerWindow.cpp

```diff
@@ -0,0 +1,222 @@
+
+    // Ask tools to generate draw commands
+    forEachTool([&](ManipulationTool* tool) { tool->draw(sceneView, gesture); });
+  }
+
+  // Manipulators may have changed the scene.
+  // Update state, build BVH, etc...
+  //[...]
+}
+
+void
+ViewerWindow::redraw()
+{
+  m_clock.tick();
+  // m_gesture.setTimeIncrement(m_clock.timeIncrement);
+  // Display frame rate in window title
+  double interval = m_clock.time - m_lastTimeCheck; //< Interval in seconds
+  m_increments += 1;
+  // update once per second
+  if (interval >= 1.0) {
+    // Compute average frame rate over the last second, if different than what we
+    // display previously, update the window title.
+    double newFrameRate = round(m_increments / interval);
+    if (m_frameRate != newFrameRate) {
+      m_frameRate = newFrameRate;
+
+      // TODO update frame rate stats from here.
+      // char title[256];
+      // snprintf(title, 256, "%s | %d fps", windowTitle, frameRate);
+      // glfwSetWindowTitle(mainWindow.handle, title);
+    }
+    m_lastTimeCheck = m_clock.time;
+    m_increments = 0;
+  }
+
+  glm::ivec2 oldpickbuffersize = m_selection.resolution;
+  bool ok = m_selection.update(glm::ivec2(width(), height()));
+  if (!ok) {
+    LOG_ERROR << "Failed to update selection buffer";
+  }
+
+  // renderer size may have been directly manipulated by e.g. the renderdialog
+  uint32_t oldrendererwidth, oldrendererheight;
+  m_renderer->getSize(oldrendererwidth, oldrendererheight);
+  if (width() != oldpickbuffersize.x || height() != oldpickbuffersize.y || width() != oldrendererwidth ||
+      height() != oldrendererheight) {
+    m_renderer->resize(width(), height());
+    m_CCamera.m_Film.m_Resolution.SetResX(width());
+    m_CCamera.m_Film.m_Resolution.SetResY(height());
+  }
+
+  sceneView.viewport.region = { { 0, 0 }, { width(), height() } };
+  sceneView.camera = m_CCamera;
+  sceneView.scene = m_renderer->scene();
+  sceneView.renderSettings = m_renderSettings;
+
+  update(sceneView.viewport, m_clock, gesture);
+  sceneView.camera.Update();
+
+  m_renderer->render(sceneView.camera);
+
+  gesture.graphics.draw(sceneView, m_selection);
+
+  // Make sure we consumed any unused input event before we poll new events.
+  // (in the case of Qt we are not explicitly polling but using signals/slots.)
+  gesture.input.consume();
+}
+
+void
+ViewerWindow::setRenderer(int rendererType)
+{
+  // clean up old renderer.
+  if (m_renderer) {
+    m_renderer->cleanUpResources();
+  }
+
+  Scene* sc = m_renderer->scene();
+
+  switch (rendererType) {
+    case 1:
+      LOG_DEBUG << "Set OpenGL pathtrace Renderer";
+      m_renderer.reset(new RenderGLPT(m_renderSettings));
+      m_renderSettings->m_DirtyFlags.SetFlag(TransferFunctionDirty);
+      break;
+    case 2:
+      LOG_DEBUG << "Set OpenGL pathtrace Renderer";
+      m_renderer.reset(new RenderGLPT(m_renderSettings));
+      m_renderSettings->m_DirtyFlags.SetFlag(TransferFunctionDirty);
+      break;
```

Comment:
Are both these cases intentionally the same?

Reply:
Yes intentional, I think there is some legacy reason for it where "2" used to be valid.

Reply:
Would it be better to merge the two cases then?
```
case 1:
case 2:
  <code here>
```